Add Python 3.3+ and Windows support

tcalmant opened the pull request:
Hi,

I've added support for Python 3.3+ for all platforms and support for the Windows platform for Pyleus.
It has been tested with the unit tests and the word count example.

This contribution is supported by Tec21 and ERODS team of the Université Grenoble Alpes.

Cheers,
Thomas

tcalmant committed:
Added a Windows's version of buildCommand

The OS is detected in the buildCommand method.
The previous buildCommand method has been renamed buildCommandUnix.

tcalmant committed:
Use StringUtils.join to make OS agnostic path names

tcalmant committed:
Use File.separator instead of System.getProperty(...)

tcalmant committed:
setup.py: call Maven directly if make is missing

Support Windows when calling setup.py.

tcalmant committed:
Add support for wheel generation

Import the .jar file in the wheel

tcalmant committed:
Register the bdist_wheel customization

tcalmant committed:
Use entry_points instead of scripts in setup.py

This will generate a pyleus.exe file on Windows, usable in the console.

tcalmant committed:
New venv_proxy, for Python 3.3+

Replaces virtualenv_proxy for Python 3.3+

tcalmant committed:
Use the venv_proxy when venv is available

... else, fall back on virtualenv.

tcalmant committed:
Correct the regular expression in cli.build._zip_dir

... according to tests

tcalmant committed:
Corrected tests to use OS-independent paths

Replaced "foo/bar"-style constant paths by calls to os.path.join()

tcalmant committed:
Reuse the VirtualenvError instead of OSError in venv_proxy

tcalmant committed:
Skip virtualenv test on Python 3.3+

tcalmant committed:
Added tests for venv_proxy

Tests on the creation of the virtual environment are missing, because it doesn't call a command anymore but uses the venv module directly

tcalmant committed:
Import print_function from __future__ where print is used.

tcalmant committed:
Correction of the regex generation in _zip_dir

Windows paths contains '\' characters, which must be escaped.'

tcalmant committed:
word_count example: write logs in the gettempdir() folder

tcalmant committed:
PyleusTopologyBuilder constants are in-jar paths

... '/' must be used instead of File.separator

tcalmant committed:
Use pytest.skip to avoid failing venv tests on an ImportError

pytestmark is handled after @mock decorators are executed, but the latters fail if venv_proxy fails to load.

tcalmant committed:
shutil.which didn't exist before Python 3.3

Use it when available, else use the command name on Posix or consider it unavailable on Windows

tcalmant committed:
Use cmd to run the Python interpreter on Windows

This seems to avoid the "file not found" bug

tcalmant committed:
Ensure we use stdin and stdout in binary mode

tcalmant committed:
Force encoding to UTF-8 in msgpack serializer

This avoids having dictionary keys in binary instead of string in Python 3

poros commented:
Hi,

Travis is failing. Could you please check the output out and fix whatever is causing the error?

To be honest, I thought we already were Python 3 ready (at least we had some contributions to make this happen) but at the end we blocked on Storm incompatibility with Python 3 (see comments in #50).

tcalmant committed:
Correction for pyflakes (unused import in tests)

tcalmant commented:
The error to come from flakes in a test to check if venv is present; I'm working on it.

The "develop" branch had some Python 3-aware code but I didn't manage to make it work as is.

tcalmant commented:
Oh, and the "test_read_msg_dict" test error is due to the behaviour of msgpack, forcing it to give back string in all cases to avoid backward incompatibility in the main.
The test using explicit "bytes" strings, is this to test the conservation of types or just for the Python 3 compatibility ?

tcalmant committed:
Updated tests to work only with strings

tcalmant committed:
Updated all examples to work on Windows

tcalmant committed:
Added support for mixed operating systems

- Allows to run the Nimbus server on Windows and components on Linux (for example)
- Fixes #2

tcalmant committed:
Call pip using "python -m"

This avoids some errors when using the "system_site_packages" option of Pyleus, where the pip package is not installed (already installed in system site packages)

westover commented:
Is there anything I can do to help with this endeavor?

tcalmant commented:
I just have some commits to cherry-pick from my other branches to support calls to "pip" on Windows while using system-site packages.
It would be nice to have some feedback from other Python 3/Windows users to look for bugs in some cases.

The version of Pyleus I use is based on a modified version of MsgPack-Java that uses Latin-1 to transfer binary data instead of UTF-8, due to some issues transferring Numpy arrays.

westover commented:
So Travis is dead now how can we revive this?
